src/workload.py

```diff
@@ -0,0 +1,238 @@
+        if service is None:
+            return False
+        return service.startup == "enabled"
+
+    @property
+    def version(self) -> str:
+        """MySQL Router version"""
+        process = self._container.exec(["mysqlrouter", "--version"])
+        raw_version, _ = process.wait_output()
+        for version in raw_version.split():
+            if version.startswith("8"):
+                return version
+        return ""
+
+
+@dataclasses.dataclass(kw_only=True)
+class AuthenticatedWorkload(Workload):
+    """Workload with connection to MySQL cluster"""
+
+    _admin_username: str
+    _admin_password: str
+    _host: str
+    _port: str
+    _charm: "charm.MySQLRouterOperatorCharm"
+
+    _UNIX_USERNAME = "mysql"
+    _ROUTER_CONFIG_DIRECTORY = pathlib.Path("/etc/mysqlrouter")
+    _ROUTER_CONFIG_FILE = "mysqlrouter.conf"
+    _TLS_CONFIG_FILE = "tls.conf"
+    _TLS_KEY_FILE = "custom-key.pem"
+    _TLS_CERTIFICATE_FILE = "custom-certificate.pem"
+
+    @property
+    def _router_username(self) -> str:
+        unit_id = self._charm.unit.name.split("/")[1]
+        return f"mysqlrouter_{unit_id}"
+
+    def _update_layer(self, *, enabled: bool, tls: bool = None) -> None:
+        """Update and restart services.
+
+        Args:
+            enabled: Whether MySQL Router service is enabled
+            tls: Whether TLS is enabled. Required if enabled=True
+        """
+        if enabled:
+            command = (
+                f"mysqlrouter --config {self._ROUTER_CONFIG_DIRECTORY / self._ROUTER_CONFIG_FILE}"
+            )
+            assert tls is not None, "`tls` argument required when enabled=True"
+            if tls:
+                command = f"{command} --extra-config {self._ROUTER_CONFIG_DIRECTORY / self._TLS_CONFIG_FILE}"
+        else:
+            command = ""
+        layer = ops.pebble.Layer(
+            {
+                "summary": "mysql router layer",
+                "description": "the pebble config layer for mysql router",
+                "services": {
+                    self._SERVICE_NAME: {
+                        "override": "replace",
+                        "summary": "mysql router",
+                        "command": command,
+                        "startup": "enabled" if enabled else "disabled",
+                        "user": self._UNIX_USERNAME,
+                        "group": self._UNIX_USERNAME,
+                    },
+                },
+            }
+        )
+        self._container.add_layer(self._SERVICE_NAME, layer, combine=True)
+        self._container.replan()
+
+    @property
+    def shell(self) -> mysql_shell.Shell:
+        """MySQL Shell"""
+        return mysql_shell.Shell(
+            _container=self._container,
+            _username=self._admin_username,
+            _password=self._admin_password,
+            _host=self._host,
+            _port=self._port,
+        )
+
+    def _bootstrap_router(self, *, password: str, tls: bool) -> None:
```

Comment:
would we need a `force` kwarg to bootstrap multiple times if `enable` called a 2nd time (enable, then disable, then enable again)?